src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs

```diff
@@ -0,0 +1,598 @@
+                string msiPackageVersion = $"{manifestVersion}";
+
+                Log?.LogMessage($"Resolving {manifestId} ({manifestVersion}) to {msiPackageId}.{msiPackageVersion}.");
+
+                // Retrieve the payload from the MSI package cache. 
+                MsiPayload msiPayload = GetCachedMsiPayload(msiPackageId, msiPackageVersion);
+                VerifyPackage(msiPayload);
+            }
+            catch (Exception e)
+            {
+                LogException(e);
+                throw;
+            }
+        }
+
+        public void RepairWorkloadPack(PackInfo packInfo, SdkFeatureBand sdkFeatureBand, DirectoryPath? offlineCache = null)
+        {
+            try
+            {
+                // Determine the MSI payload package ID based on the host architecture, pack ID and pack version.
+                string msiPackageId = GetMsiPackageId(packInfo);
+
+                // Retrieve the payload from the MSI package cache. 
+                MsiPayload msiPayload = GetCachedMsiPayload(msiPackageId, packInfo.Version);
+                VerifyPackage(msiPayload);
+                DetectState state = Detect(msiPayload.Manifest.ProductCode, out string _);
+                PlanAndExecute(msiPayload, packInfo, state, InstallAction.Repair);
+
+                // Update the reference count against the MSI.
+                UpdateDependent(InstallRequestType.AddDependent, msiPayload.Manifest.ProviderKeyName, _dependent);
+            }
+            catch (Exception e)
+            {
+                LogException(e);
+                throw;
+            }
+        }
+
+        public void InstallWorkloadPack(PackInfo packInfo, SdkFeatureBand sdkFeatureBand, DirectoryPath? offlineCache = null)
+        {
+            try
+            {
+                // Determine the MSI payload package ID based on the host architecture, pack ID and pack version.
+                string msiPackageId = GetMsiPackageId(packInfo);
+
+                // Retrieve the payload from the MSI package cache. 
+                MsiPayload msiPayload = GetCachedMsiPayload(msiPackageId, packInfo.Version);
+                VerifyPackage(msiPayload);
+                DetectState state = Detect(msiPayload.Manifest.ProductCode, out string _);
+                PlanAndExecute(msiPayload, packInfo, state, InstallAction.Install);
+
+                // Update the reference count against the MSI.
+                UpdateDependent(InstallRequestType.AddDependent, msiPayload.Manifest.ProviderKeyName, _dependent);
+            }
+            catch (Exception e)
+            {
+                LogException(e);
+                throw;
+            }
+        }
+
+        public void RollBackWorkloadPackInstall(PackInfo packInfo, SdkFeatureBand sdkFeatureBand)
+        {
+            Log?.LogMessage($"Rolling back workload pack.");
+            LogPackInfo(packInfo);
```

Comment:
So nothing is actually done here besides logging?

Reply:
yup, rollbacks aren't implemented yet. I'm still mulling over how I want to track the actual actions, e.g. if we installed a workload, but we have a shared pack that was installed by another workload, a rollback probably shouldn't remove the pack, but the reference counting is done against the same feature band. Similarly we have issues with modifying registry keys where we may no-op, but then need to remember whether we actually wrote the key during the current transaction, etc.